aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -92,18 +97,111 @@ impl SessionId {
     inner: Session<'r, 'l, S>,
+    remote: MoveResolverWithVMMetadata<'r, 'l, S>,
 }
 
 impl<'r, 'l, S> SessionExt<'r, 'l, S>
 where
     S: MoveResolverExt,
 {
-    pub fn new(inner: Session<'r, 'l, S>) -> Self {
-        Self { inner }
+    pub fn new(inner: Session<'r, 'l, S>, move_vm: &'l MoveVM, remote: &'r S) -> Self {
+        Self {
+            inner,
+            remote: MoveResolverWithVMMetadata::new(remote, move_vm),
+        }
     }
 
     pub fn finish(self) -> VMResult<SessionOutput> {
         let (change_set, events, mut extensions) = self.inner.finish_with_extensions()?;
+
+        let common_error = PartialVMError::new(StatusCode::UNREACHABLE).finish(Location::Undefined);
+        let mut change_set_grouped = MoveChangeSet::new();
+        for (addr, account_changeset) in change_set.into_inner() {
+            let mut resource_groups: BTreeMap<StructTag, AccountChangeSet> = BTreeMap::new();
+            let mut account_changeset_grouped = AccountChangeSet::new();
+            let (modules, resources) = account_changeset.into_inner();
+
+            // * Separate the resource groups from the non-resource groups
+            // * non-resource groups are kept as is
+            // * resource groups are merged into the correct format as deltas to the source data
+            for (struct_tag, blob_op) in resources {
+                let resource_group = self.remote.get_resource_group(&struct_tag).map_err(|_| common_error.clone())?;
+                if let Some(resource_group) = resource_group {
+                    resource_groups
+                        .entry(resource_group)
+                        .or_insert_with(AccountChangeSet::new)
+                        .add_resource_op(struct_tag, blob_op)
+                        .map_err(|_| common_error.clone())?;
+                    continue;
+                }
+
+                account_changeset_grouped
+                    .add_resource_op(struct_tag, blob_op)
+                    .map_err(|_| common_error.clone())?;
+            }
+
+            for (resource_tag, resources) in resource_groups {
+                let source_data = self
+                    .remote
+                    .get_resource_group_data(&addr, &resource_tag)
+                    .map_err(|_| common_error.clone())?;
+                let (mut source_data, create) = if let Some(source_data) = source_data {
+                    let source_data =
+                        bcs::from_bytes(&source_data).map_err(|_| common_error.clone())?;
+                    (source_data, false)
+                } else {
+                    (BTreeMap::new(), true)
+                };
+
+                for (struct_tag, current_op) in resources.into_resources() {
+                    match current_op {
+                        MoveStorageOp::Delete => {
+                            source_data
+                                .remove(&struct_tag)
+                                .ok_or_else(|| common_error.clone())?;
```

Comment:
no, again, if we hit this, it is likely due to some broken invariant.

why are we getting a call to delete something from the VM if it doesn't exist?

Reply:
Why not INTERNAL_INVARIANT_VIOLATION then?

Reply:
I guess I'm not sure when to use unreachable versus invariant violation. They seem to be similar. Perhaps one is used when the logic is self contained versus the other when the logic might be more distributed?

Reply:
Note, there's no such status code. I'm okay with unreachable but willing to change with sufficient motivation.

Reply:
It's StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR. The reason I brought this up is because we might have metrics/monitoring specifically monitor this error code but not UNREACHABLE. cc @wrwg as well

Reply:
done